File: packages/typescript/goldenmatch/src/core/lshBlocker.ts
/**
 * lshBlocker.ts — MinHash/LSH near-duplicate blocking (#1081).
 *
 * Edge-safe: no Node.js imports. Pure-TS port of the Python
 * `core/lsh_blocker.py::MinHashLSHBlocker`, over an in-memory `string[]` of
 * record texts (the Python sibling is Polars-coupled; this TS utility keeps the
 * same bucketing/pairing semantics without a DataFrame layer).
 *
 * Shingle each text, MinHash it, bucket the signature into bands, and group
 * records by `(band_idx, bucket)`. Records sharing >= 1 bucket are candidates; a
 * pair colliding in several bands is de-duplicated via a canonical `"min,max"`
 * pair key. Empty / whitespace-only texts have nothing to block on (their
 * signature is all-u64::MAX) and are dropped by comparing against a
 * deterministic sentinel.
 */

import { optimalBands, sketchBandHashes } from "./sketch.js";

/** Resolved MinHash/LSH parameters. Mirrors Python `LSHKeyConfig` (resolved). */
export interface MinHashLSHConfig {
  /** Shingle granularity: `"char"` (code points) or `"word"` (tokens). */
  readonly mode: string;
  /** Shingle size (k-gram). Must be >= 1. */
  readonly k: number;
  /** Number of MinHash permutations (signature length). */
  readonly numPerms: number;
  /**
   * Number of LSH bands. Provide this OR `threshold` (resolved via
   * `optimalBands`). `numPerms` must be divisible by `numBands`.
   */
  readonly numBands?: number;
  /** Similarity threshold that drives `optimalBands` when `numBands` is absent. */
  readonly threshold?: number;
  /** splitmix64 seed for the permutation coefficients. */
  readonly seed: bigint;
}

/** Canonical `(min, max)` pair key, matching the project-wide invariant. */
function canonicalPairKey(a: number, b: number): string {
  return a < b ? `${a},${b}` : `${b},${a}`;
}

export class MinHashLSHBlocker {
  readonly mode: string;
  readonly k: number;
  readonly numPerms: number;
  readonly numBands: number;
  readonly seed: bigint;

  constructor(
    mode: string,
    k: number,
    numPerms: number,
    numBands: number,
    seed: bigint,
  ) {
    this.mode = mode;
    this.k = k;
    this.numPerms = numPerms;
    this.numBands = numBands;
    this.seed = seed;
  }

  /**
   * Construct from a (possibly threshold-driven) config. When `numBands` is
   * given it is used directly; otherwise `optimalBands(numPerms, threshold)`
   * picks the `(b, r)` split.
   */
  static fromConfig(cfg: MinHashLSHConfig): MinHashLSHBlocker {
    let numBands: number;
    if (cfg.numBands !== undefined) {
      numBands = cfg.numBands;
    } else {
      if (cfg.threshold === undefined) {
        throw new Error(
          "MinHashLSHBlocker.fromConfig requires either numBands or threshold",
        );
      }
      [numBands] = optimalBands(cfg.numPerms, cfg.threshold);
    }
    return new MinHashLSHBlocker(cfg.mode, cfg.k, cfg.numPerms, numBands, cfg.seed);
  }

  /**
   * Band hashes of an empty record (all-u64::MAX signature). Empty /
   * whitespace-only texts all produce this deterministic sentinel; we detect and
   * drop them by comparison (a non-empty record cannot produce an all-MAX
   * signature, so the comparison is exact).
   */
  private emptySentinel(): bigint[] {
    return sketchBandHashes(
      "",
      this.mode,
      this.k,
      this.numPerms,
      this.numBands,
      this.seed,
    );
  }

  /**
   * Map `"band_idx,bucket"` -> row positions, skipping empty / whitespace-only
   * rows. The key encodes the band index so identical bucket hashes in
   * different bands stay separate groups.
   */
  buckets(texts: readonly string[]): Map<string, number[]> {
    const sentinel = this.emptySentinel();
    const sentinelKey = sentinel.join(",");
    const groups = new Map<string, number[]>();

    for (let rowIdx = 0; rowIdx < texts.length; rowIdx++) {
      const bands = sketchBandHashes(
        texts[rowIdx]!,
        this.mode,
        this.k,
        this.numPerms,
        this.numBands,
        this.seed,
      );
      // empty / whitespace-only: no content to block on.
      if (bands.join(",") === sentinelKey) continue;
      for (let bandIdx = 0; bandIdx < bands.length; bandIdx++) {
        const key = `${bandIdx},${bands[bandIdx]!}`;
        const existing = groups.get(key);
        if (existing !== undefined) existing.push(rowIdx);
        else groups.set(key, [rowIdx]);
      }
    }
    return groups;
  }

  /**
   * De-duplicated `(min, max)` candidate pairs across all bands, as a set of
   * canonical `"min,max"` string keys. A pair colliding in several bands is
   * offered once.
   */
  candidatePairs(texts: readonly string[]): Set<string> {
    const pairs = new Set<string>();
    for (const members of this.buckets(texts).values()) {
      if (members.length < 2) continue;
      for (let i = 0; i < members.length; i++) {
        for (let j = i + 1; j < members.length; j++) {
          pairs.add(canonicalPairKey(members[i]!, members[j]!));
        }
      }
    }
    return pairs;
  }
}

File: packages/python/goldenmatch/goldenmatch/core/lsh_blocker.py
"""MinHash/LSH blocking (#1081).

A near-duplicate candidate-generation blocker built on the sketch kernel
(``core/sketch.py``): shingle each record's text column, MinHash it, bucket the
signature into bands, and emit one block per non-singleton ``(band, bucket)``
group. Records sharing >= 1 bucket are candidates; a pair colliding in several
bands is scored once thanks to the downstream ``(min, max)`` pair de-dup (or use
``candidate_pairs`` for a de-duplicated pair set directly).

Conforms to the existing ``BlockResult`` blocker contract; ``blocker.build_blocks``
dispatches here for ``strategy="lsh"``.
"""
from __future__ import annotations

from dataclasses import dataclass

import polars as pl

from goldenmatch.config.schemas import BlockingConfig, LSHKeyConfig
from goldenmatch.core import sketch


@dataclass
class MinHashLSHBlocker:
    """Resolved MinHash/LSH parameters + the bucketing/blocking operations."""

    mode: str
    k: int
    num_perms: int
    num_bands: int
    seed: int

    @classmethod
    def from_config(cls, lsh: LSHKeyConfig) -> MinHashLSHBlocker:
        if lsh.num_bands is not None:
            num_bands = lsh.num_bands
        else:
            # threshold-driven: optimal_bands picks the (b, r) split.
            assert lsh.threshold is not None  # guaranteed by LSHKeyConfig validation
            num_bands, _ = sketch.optimal_bands(lsh.num_perms, lsh.threshold)
        return cls(lsh.mode, lsh.k, lsh.num_perms, num_bands, lsh.seed)

    def _empty_sentinel(self) -> list[int]:
        """Band hashes of an empty record (all-``u64::MAX`` signature).

        Empty / whitespace-only texts have nothing to block on; they would
        otherwise all collide into one giant block. We detect and drop them by
        comparing against this deterministic sentinel (a non-empty record cannot
        produce an all-MAX signature, so the comparison is exact).
        """
        return sketch.sketch_band_hashes(
            "", self.mode, self.k, self.num_perms, self.num_bands, self.seed
        )

    def buckets(self, texts: list[str]) -> dict[tuple[int, int], list[int]]:
        """Map ``(band_idx, bucket_hash)`` -> row positions, skipping empty rows."""
        band_hashes = sketch.band_hashes_batch(
            texts, self.mode, self.k, self.num_perms, self.num_bands, self.seed
        )
        sentinel = self._empty_sentinel()
        groups: dict[tuple[int, int], list[int]] = {}
        for row_idx, bands in enumerate(band_hashes):
            if bands == sentinel:  # empty / whitespace-only: no content to block on
                continue
            for band_idx, bucket in enumerate(bands):
                groups.setdefault((band_idx, bucket), []).append(row_idx)
        return groups

    def candidate_pairs(self, texts: list[str]) -> set[tuple[int, int]]:
        """De-duplicated ``(min, max)`` candidate pairs across all bands."""
        pairs: set[tuple[int, int]] = set()
        for members in self.buckets(texts).values():
            if len(members) < 2:
                continue
            for i in range(len(members)):
                for j in range(i + 1, len(members)):
                    a, b = members[i], members[j]
                    pairs.add((a, b) if a < b else (b, a))
        return pairs

    def blocks(self, df: pl.DataFrame, texts: list[str]) -> list:
        """One ``BlockResult`` per non-singleton ``(band, bucket)`` group."""
        from goldenmatch.core.blocker import BlockResult

        results: list[BlockResult] = []
        for (band_idx, bucket), members in self.buckets(texts).items():
            if len(members) < 2:
                continue
            block_df = df[members]  # positional row select (preserves __row_id__)
            results.append(
                BlockResult(
                    block_key=f"lsh_b{band_idx}_{bucket}",
                    df=block_df.lazy(),
                    strategy="minhash_lsh",
                )
            )
        return results


def build_lsh_blocks(lf: pl.LazyFrame, config: BlockingConfig) -> list:
    """Build MinHash/LSH blocks for ``strategy="lsh"`` (called by ``build_blocks``)."""
    if config.lsh is None:
        raise ValueError("LSH blocking requires a 'lsh' config block.")
    df = lf.collect()
    if config.lsh.column not in df.columns:
        raise ValueError(f"LSH blocking column {config.lsh.column!r} not found in data.")
    texts = df[config.lsh.column].cast(pl.Utf8).fill_null("").to_list()
    return MinHashLSHBlocker.from_config(config.lsh).blocks(df, texts)
